ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,259 @@
-                if (NT_SUCCESS(Status))
-                {
-                    if (DeviceObject == EntryDeviceObject)
-                    {
-                        if (Event == &GUID_PNP_CUSTOM_NOTIFICATION)
-                        {
-                            ((PTARGET_DEVICE_CUSTOM_NOTIFICATION)NotificationStructure)->FileObject = ChangeEntry->FileObject;
-                        }
-                        else
-                        {
-                            ((PTARGET_DEVICE_REMOVAL_NOTIFICATION)NotificationStructure)->FileObject = ChangeEntry->FileObject;
-                        }
-                        CallCurrentEntry = TRUE;
-                    }
-                }
-                break;
-            }
-            default:
-            {
-                DPRINT1("IopNotifyPlugPlayNotification(): unknown EventCategory 0x%x UNIMPLEMENTED\n", EventCategory);
-                break;
-            }
-        }
+        PipProcessSingleNotification(ChangeEntry, EventCategory, NotificationStruct, DeviceObject);
+    }
+
+    // Now process added elements (if any)
+    ListEntry = PiAddedNotifyListHead.Flink;
+    while ((ListEntry = ExInterlockedRemoveHeadList(&PiAddedNotifyListHead, &PiNotifyAdditionLock)))
+    {
+        // append the entry to PiNotifyListHead while we're holding PiNotificationInProgressLock
+        InsertHeadList(&PiNotifyListHead, ListEntry);
+
+        ChangeEntry = CONTAINING_RECORD(ListEntry, PNP_NOTIFY_ENTRY, PnpNotifyList);
+
+        PipProcessSingleNotification(ChangeEntry, EventCategory, NotificationStruct, DeviceObject);
+    }
+
+    KeReleaseGuardedMutex(&PiNotificationInProgressLock);
 
-        /* Move to the next element now, as callback may unregister itself */
-        ListEntry = ListEntry->Flink;
-        /* FIXME: If ListEntry was the last element and that callback registers
-         * new notifications, those won't be checked... */
+    ExFreePoolWithTag(NotificationStruct, TAG_PNP_NOTIFY);
+}
+
+/**
+ * @brief      Delivers the event to all drivers subscribed to
+ *             EventCategoryDeviceInterfaceChange
+ *
+ * @param[in]  Event               The PnP event GUID
+ * @param[in]  InterfaceClassGuid  The GUID of an interface class
+ * @param[in]  SymbolicLinkName    Pointer to a string identifying the device interface name
+ */
+VOID
+PiNotifyDeviceInterfaceChange(
+    _In_ LPCGUID Event,
+    _In_ LPCGUID InterfaceClassGuid,
+    _In_ PUNICODE_STRING SymbolicLinkName)
+{
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION notifStruct;
+    notifStruct = ExAllocatePoolWithTag(PagedPool, sizeof(*notifStruct), TAG_PNP_NOTIFY);
+    if (!notifStruct)
+    {
+        return;
+    }
 
-        if (CallCurrentEntry)
+    *notifStruct = (DEVICE_INTERFACE_CHANGE_NOTIFICATION) {
+        .Version = 1,
+        .Size = sizeof(DEVICE_INTERFACE_CHANGE_NOTIFICATION),
+        .Event = *Event,
+        .InterfaceClassGuid = *InterfaceClassGuid,
+        .SymbolicLinkName = SymbolicLinkName
+    };
```

Comment:
That's designated initializers from C99. Pretty standard, and supported e.g. with MSVC 2013+.

Reply:
Please don't use that in the kerne and keep the current code stylel; it should remain C89-compatible.

Reply:
Maybe this will immediately solve all the problems? :)

/* Windows NT target version 10.19H1 */
#define VER_PRODUCTMAJORVERSION  10
#define VER_PRODUCTMINORVERSION   7

add_definitions(-DWINVER=0xA00
                        -D_WIN32_IE=0xA00
                        -D_WIN32_WINNT=0xA00
                        -D_WIN32_WINDOWS=0xA00
                        -D_SETUPAPI_VER=0xA00
                        -DMINGW_HAS_SECURE_API=1)